src/components/SignIn/SignInForm.js

```diff
@@ -0,0 +1,146 @@
+import Button from '@material-ui/core/Button';
+import CircularProgress from '@material-ui/core/CircularProgress';
+import FormControl from '@material-ui/core/FormControl';
+import Grid from '@material-ui/core/Grid';
+import Link from '@material-ui/core/Link';
+import { makeStyles, withStyles } from '@material-ui/core/styles';
+import Container from '@material-ui/core/Container';
+import TextField from '@material-ui/core/TextField';
+import Typography from '@material-ui/core/Typography';
+import firebase from 'firebase/app';
+import 'firebase/auth';
+import React, { useCallback, useMemo, useState } from 'react';
+import PropTypes from 'prop-types';
+import * as ROUTES from '../../constants/routes';
+
+const RoundedTextField = withStyles((theme) => ({
+  root: {
+    '& fieldset': {
+      borderRadius: 30,
+    },
+    '& input': {
+      marginLeft: theme.spacing(1),
+      marginRight: theme.spacing(1),
+    },
+  },
+}))(TextField);
+
+const RoundedButton = withStyles((theme) => ({
+  root: {
+    borderRadius: 30,
+    height: theme.typography.fontSize * 3,
+    margin: theme.spacing(3, 0, 2),
+    textTransform: 'none',
+    width: '50%',
+  },
+}))(Button);
+
+const useStyles = makeStyles({
+  container: {
+    display: 'flex',
+    flexDirection: 'column',
+    alignItems: 'center',
+  },
+  buttonWrapper: {
+    display: 'flex',
+    flexDirection: 'column',
+    alignItems: 'center',
+    position: 'relative',
+  },
+  buttonProgress: {
+    position: 'absolute',
+    top: '50%',
+    left: '50%',
+    marginTop: -8,
+    marginLeft: -8,
+  },
+  linksWrapper: {
+    marginTop: 15,
+  },
+  welcome: {
+    marginBottom: 30,
+  },
+});
+
+export default function SignInForm({ setError }) {
+  const classes = useStyles();
+
+  const [email, setEmail] = useState('');
+  const [password, setPassword] = useState('');
+  const [loading, setLoading] = useState(false);
+
+  const handleSubmit = useCallback(() => {
+    setLoading(true);
+    firebase.auth().signInWithEmailAndPassword(email, password)
+      .catch((error) => {
+        setError(error);
+      })
+      .finally(() => {
+        setLoading(false);
+      });
+  }, [email, password, setError]);
+
+  const isInvalid = useMemo(() => password === '' || email === '',
+    [email, password]);
+
+  return (
+    <Container className={classes.container}>
+      <Typography variant="h4" className={classes.welcome}>Welcome back</Typography>
+      <FormControl>
+        <RoundedTextField
+          autoFocus
+          variant="outlined"
+          margin="normal"
+          required
+          fullWidth
+          id="email"
+          label="Email"
+          name="email"
+          autoComplete="email"
+          value={email}
```

Comment:
That's a lot of "email"s, maybe use an array of `["email", "password"]` or something similar and map the array to `RoundedTextField` to reduce code reuse?

Reply:
Actually now that i'm looking at it closer, the `id` and `name` props aren't needed. So that should cut it down to just `email`, `"Email"`, and `"email"`. Array mapping sounds good!